src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Custom/Serializer/ActivitySerializer.cs

```diff
@@ -0,0 +1,428 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using OpenTelemetry.Internal;
+using OpenTelemetry.Resources;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.Custom.Serializer;
+
+internal class ActivitySerializer
+{
+    private static readonly ConcurrentBag<List<Activity>> ActivityListPool = new();
+
+    private readonly SdkLimitOptions sdkLimitOptions;
+
+    private readonly ActivitySizeCalculator activitySizeCalculator;
+
+    internal ActivitySerializer(SdkLimitOptions sdkLimitOptions)
+    {
+        this.sdkLimitOptions = sdkLimitOptions;
+        this.activitySizeCalculator = new ActivitySizeCalculator(sdkLimitOptions);
+    }
+
+    internal int Serialize(ref byte[] buffer, int offset, Resource? resource, Batch<Activity> batch)
+    {
+        Dictionary<string, List<Activity>> scopeTraces = new();
+        foreach (var activity in batch)
+        {
+            if (scopeTraces.TryGetValue(activity.Source.Name, out var activityList))
+            {
+                activityList.Add(activity);
+            }
+            else
+            {
+                if (!ActivityListPool.TryTake(out var newList))
```

Comment:
Do we need such complex storage? I understand that we need to track a single `ActivitySource.Name` for the instrumentation scope. Can we rely on `Activity.Source` and a simple thing like `isSeen` during the serialization process?